### ISSUE TITLE
Name of context and name of cluster should match name given in config

### ISSUE THREAD

**Issue by jpmcb** [kubernetes-sigs/kind#2668]
Currently, given the following config:

```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
name: my-test-cluster
```

The cluster that is created has the name `kind-my-test-cluster` and the context `kind-my-test-cluster`. Note that `kind-` is prefixed.

This can be seen via kubectl:
```
$  kubectl config view --minify
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: DATA+OMITTED
    server: https://127.0.0.1:33597
  name: kind-my-test-cluster
contexts:
- context:
    cluster: kind-my-test-cluster
    user: kind-my-test-cluster
  name: kind-my-test-cluster
current-context: kind-my-test-cluster
kind: Config
preferences: {}
users:
- name: kind-my-test-cluster
  user:
    client-certificate-data: REDACTED
    client-key-data: REDACTED
```

But the `kind-` prefix is omitted when listing clusters:
```
$ kind get clusters
my-test-cluster
```

and the nodes also get named _without_ the prefix:
```
$ kubectl get nodes
NAME                            STATUS   ROLES                  AGE    VERSION
my-test-cluster-control-plane   Ready    control-plane,master   3m2s   v1.21.1
```

### The problem

This (in my opinion) is a pretty confusing user experience and makes creating automation or programs difficult without adding the `kind-` prefix when targeting a cluster or context. A user might not know about the `kind-` prefix to the name of their cluster and in many regards, it's hidden.

Further, when using the `kind` Go libraries, the name provided in the `Create` command is used, but has `kind-` prefixed:
```go
package main

import (
        kindcluster "sigs.k8s.io/kind/pkg/cluster"
)

func main() {
        kindProvider := kindcluster.NewProvider()
        clusterConfig := kindcluster.CreateWithKubeconfigPath("/tmp/my-kind.yaml")

        // Give a specific name of the kind cluster to create.
        // The _actual_ name of the cluster will be "kind-my-cluster"
        err := kindProvider.Create("my-cluster", clusterConfig)
        if err != nil {
                panic(err)
        }
}

```

After running the above go program, the cluster is created and is listable by `kind`:
```
$ kind get clusters
my-cluster
```

but looking at it's kubeconfig yaml, it is named `kind-my-cluster`:
```
$ head /tmp/my-cluster.yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: ...
    server: https://127.0.0.1:38971
  name: kind-my-cluster
contexts:
- context:
    cluster: kind-my-cluster
    user: kind-my-cluster
```

_**As a consumer of the `kind` go APIs, I would have expected the cluster and the context to be set as `my-cluster`, the name that I have provided**_ 

My assumption is this was for historical reasons and changing this default behavior is not easy? Is there a way to programmatically overwrite this? Do the `kind` libraries provide a way to change this so the name matches across different areas?

**Comment by BenTheElder** [comment]
> But the kind- prefix is omitted when listing clusters:

Yes, because the cluster name is `my-test-cluster`

Kubeconfig context are typically prefixed by tool in order to avoid naming collisions / editing config entries created by some other cluster management tool.

GKE for example does `gke_${name}` in KUBECONFIG IIRC. k3d does similar

> This (in my opinion) is a pretty confusing user experience and makes creating automation or programs difficult without adding the kind- prefix when targeting a cluster or context. A user might not know about the kind- prefix to the name of their cluster and in many regards, it's hidden.

I disagree that it is hidden, the output of `kind create cluster` gives a sample kubectl command with the correct `--context`.

We also discuss this https://kind.sigs.k8s.io/docs/user/quick-start/#interacting-with-your-cluster


> My assumption is this was for historical reasons and changing this default behavior is not easy?

No, it's very much intentional. It avoids naming collisions and makes kind-based clusters readily identifiable to various tools like tilt / skaffold.

> Is there a way to programmatically overwrite this? 

No, but you can always edit the kubeconfig entries with some other tooling including kubectl, the names just need to be consistent within the kubeconfig because the entries reference each other.

`kubectl config rename-context old-name new-name`

> Do the kind libraries provide a way to change this so the name matches across different areas?

No. We avoid unnecessary branching behavior. This can trivially be done with post-processing.


See also https://github.com/k3d-io/k3d/issues/387

**Comment by stmcginnis** [comment]
Shouldn't the *CLUSTER* column have the matching cluster name?

```
$ k config get-contexts 
CURRENT   NAME                            CLUSTER       AUTHINFO            NAMESPACE
          kind-foo2                       kind-foo2     kind-foo2
```

**Comment by BenTheElder** [comment]
no, all kubeconfig entries are prefixed to avoid collisions e.g.

```
CURRENT   NAME                                                         CLUSTER                                                      AUTHINFO                                                     NAMESPACE
          gke_k8s-prow-builds_us-central1-f_prow                       gke_k8s-prow-builds_us-central1-f_prow                       gke_k8s-prow-builds_us-central1-f_prow                                         
```

**Comment by jpmcb** [comment]
Sounds good - this was partly a ❓ and partly 🦆, so thanks for the very thoughtful insights.

> Kubeconfig context are typically prefixed by tool in order to avoid naming collisions 

That's good to know (and isn't really a paradigm we've investigated)

---

So, another question: if we are using the `kind` Go libraries to bootstrap a new cluster and want a separate, _generic_ tool to be able to interact with that cluster by name, maybe the best way to have it all wrapped up and succinct would be to rename things within the kubeconfig?

In other words, as a very simple example, if I bootstrap a cluster with:
```
Alexa, bootstrap a local cluster named my-cluster
```
and it creates the cluster/context named `kind-my-cluster`, if I want to then do:
```
Hey Google, give me diagnostics data on my-cluster
```
The best way to go about this would be to maybe rewrite the name and context within the kubeconfig bits?

**Comment by BenTheElder** [comment]
> [...] if we are using the kind Go libraries to bootstrap a new cluster and want a separate, generic tool to be able to interact with that cluster by name, maybe the best way to have it all wrapped up and succinct would be to rename things within the kubeconfig?
>
> [...]
>
> The best way to go about this would be to maybe rewrite the name and context within the kubeconfig bits?

Yes, if you're using the kind libraries you can also e.g. export the kubeconfig in memory and manipulate it there as well, control where the config is exported to etc.

If you need to do config merging / manipulation, client-go is usually a good choice and has facilities for this. KIND admittedly doesn't use client-go for this, because we need to be independent of Kubernetes versions / chicken and egg testing issue 🙃

**Comment by stmcginnis** [comment]
@jpmcb is anything more needed here? Or can this issue be closed out?

**Comment by jorgemoralespou** [comment]
@jpmcb be aware that Kind removes the Kubeconfig entries on cluster deletion, so that if you rename the context, these will not be removed by kind deletion process, so you will also to create a `deletion` action to delete the post-processed entries.

**Comment by sathieu** [comment]
You can also use `kubectl config rename-context kind-foo2 foo2`
